src/portfolio_solve.cpp

```diff
@@ -0,0 +1,80 @@
+/*********************                                                        */
+/*! \file portfolio_solver.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Amalee Wilson
+** This file is part of the smt-switch project.
+** Copyright (c) 2021 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Implementation of a portfolio solving function that takes a vector
+**        of solvers and a term, and returns check_sat from the first solver
+**        that finishes.
+**/
+
+#include "portfolio_solve.h"
+
+namespace smt {
+
+// Used for synchronization.
+std::mutex m;
+std::condition_variable cv;
+
+// Once a solver is done, is_sat has been set,
+// and the main thread can terminate the others.
+bool a_solver_is_done = false;
```

Comment:
This might just be me not understanding the multi-threading part, but why is this variable outside the function? What if you call `portfolio_solve` twice in a row? Is it reset when `portfolio_solve` completes?

Reply:
I failed to imagine a case where someone would call portfolio_solve twice in a row, but I can make it so that it handles that. 

The variable is outside the function because all threads that execute the function `do_solver` (and the main thread executing `portfolio_solve`) are sharing this variable. Once a thread has finished, it acquires the mutex and marks a_solver_is_done as true. The condition variable in portfolio_solve is waiting for a_solver_is_done to be true. An alternative way to write this is 

```
while (!a_solver_is_done) {} // poll until true
// now cancel threads executing do_solver and return
```

Reply:
Gotcha, makes sense! I think it would be good to support multiple calls in a row if it's not too difficult. And if it is too difficult, it should throw an exception the second call. It's probably a pretty common use case to call multiple times. Even if it's not solving "incremental solving" directly, it could still be used within another framework (symbolic execution or a modeling tool) which might make many separate satisfiability checking calls in the same process.